src/lib/ngsi/MetadataVector.cpp

```diff
@@ -101,61 +102,28 @@ bool MetadataVector::matchFilter(const std::string& mdName, const std::vector<st
 *
 * If anybody needs a metadata named 'value' or 'type', then API v1
 * will have to be used to retreive that information.
+*
 */
-std::string MetadataVector::toJson(bool isLastElement, const std::vector<std::string>& metadataFilter)
+std::string MetadataVector::toJson(const std::vector<std::string>& metadataFilter)
 {
   if (vec.size() == 0)
   {
-    return "";
-  }
-
-
-  //
-  // Pass 1 - count the total number of metadatas valid for rendering.
-  //
-  // Metadatas named 'value' or 'type' are not rendered.
-  // This gives us a small problem in the logic here, about knowing whether the
-  // comma should be rendered or not.
-  //
-  // To fix this problem we need to do two passes over the vector, the first pass to
-  // count the number of valid metadatas and the second to do the work.
-  // In the second pass, if the number of rendered metadatas "so far" is less than the total
-  // number of valid metadatas, then the comma must be rendered.
-  //
-  int validMetadatas = 0;
-  for (unsigned int ix = 0; ix < vec.size(); ++ix)
-  {
-    if ((vec[ix]->name == "value") || (vec[ix]->name == "type") || !(matchFilter(vec[ix]->name, metadataFilter)))
-    {
-      continue;
-    }
-
-    ++validMetadatas;
+    return "{}";
   }
 
+  JsonHelper jh;
 
-  //
-  // And this is pass 2, where the real work is done.
-  //
-  std::string  out;
-  int          renderedMetadatas = 0;
   for (unsigned int ix = 0; ix < vec.size(); ++ix)
   {
     if ((vec[ix]->name == "value") || (vec[ix]->name == "type") || !(matchFilter(vec[ix]->name, metadataFilter)))
     {
       continue;
     }
 
-    ++renderedMetadatas;
-    out += vec[ix]->toJson(renderedMetadatas == validMetadatas);
+    jh.addRaw(vec[ix]->name, vec[ix]->toJson());
   }
 
-  if (!isLastElement)
-  {
-    out += ",";
-  }
-
-  return out;
+  return jh.str();
 }
```

Comment:
Truly great would be to have an input parameter "JsonHelper* jhP" to all these rendering functions, instead of returning objects on the stack ...

NTC, but ... perhaps in a future modification? ;-)

Reply:
You mean to have a single instance of JsonHelper per rendered tree passing it down along the try?

An interesting idea... although it should be exlored in terms of complexity  / performance gain. Perhaps in the future...

Reply:
Yes, one single JsonHelper per response. that would be good.
As you say, perhaps in the future ...
Issue?

NTC